src/runtime_src/driver/zynq/drm/zocl/zocl_sk.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * A GEM style (optionally CMA backed) device manager for ZynQ based
+ * OpenCL accelerators.
+ *
+ * Copyright (C) 2019 Xilinx, Inc. All rights reserved.
+ *
+ * Authors:
+ *    Larry Liu       <yliu@xilinx.com>
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "zocl_drv.h"
+#include "zocl_sk.h"
+#include "sched_exec.h"
+
+int
+zocl_sk_getcmd_ioctl(struct drm_device *dev, void *data, struct drm_file *filp)
+{
+	struct drm_zocl_dev *zdev = dev->dev_private;
+	struct soft_kernel *sk = zdev->soft_kernel;
+	struct soft_kernel_cmd *scmd;
+	struct drm_zocl_sk_getcmd *kdata = data;
+
+	/* If no command, the process who calls this ioctl will block here */
+	mutex_lock(&sk->sk_lock);
+	while (list_empty(&sk->sk_cmd_list)) {
+		mutex_unlock(&sk->sk_lock);
+		if (wait_event_interruptible(sk->sk_wait_queue,
+		    !list_empty(&sk->sk_cmd_list)))
+			return -ERESTARTSYS;
+		mutex_lock(&sk->sk_lock);
+	}
+
+	scmd = list_first_entry(&sk->sk_cmd_list, struct soft_kernel_cmd,
+	    skc_list);
+	list_del(&scmd->skc_list);
+	mutex_unlock(&sk->sk_lock);
+
+	kdata->opcode = scmd->skc_packet->opcode;
+
+	if (kdata->opcode == OP_CONFIG_SKERNEL) {
+		struct configure_sk_cmd *cmd;
+
+		/* Copy the command to ioctl caller */
+		cmd = (struct configure_sk_cmd *)scmd->skc_packet;
+		kdata->start_cuidx = cmd->start_cuidx;
+		kdata->cu_nums = cmd->num_cus;
+		kdata->size = cmd->sk_size;
+
+		/* soft kernel's physical address (little endian) */
+		kdata->paddr = cmd->sk_addr;
+
+		snprintf(kdata->name, ZOCL_MAX_NAME_LENGTH, "%s",
+		    (char *)cmd->sk_name);
+	} else
+		/* We will handle more opcodes */
+		DRM_WARN("Unknown soft kernel command: %d\n",
+		    kdata->opcode);
+
```

Comment:
The else part look like is simple. Should we optimized this?
if (kdata->opcode != OP_CONFIG_SKERNEL) {
         /* ... */
         DRM_WARN("Unknown soft kernel command: %d\n",
		    kdata->opcode);
         goto free_and_return;
}

struct configure_sk_cmd *cmd;
.....

free_and_return:
......

Reply:
I would rather keep it this way since this is more like a placeholder. We will add more commands type soon and switch will be a better choices.

Reply:
If this is the case, I'm okay with it.